plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs

```diff
@@ -0,0 +1,196 @@
+validateScript :: Script -> Bool
+validateScript = isRight . Flat.unflat @Scripts.Script . fromShort
+
+-- | Convenience constructor for a 'CostModel' from a map giving the parameters.
+mkCostModel :: Map String Integer -> Either EvaluationError CostModel
+-- TODO: this is a bit of a hack, replace it with something more principled
+mkCostModel m = case fromJSON $ toJSON m of
+    Success a -> pure a
+    Error e   -> throwError $ CostModelParameterMismatch e
+
+type VerboseMode = Bool
+
+type LogOutput = [Text.Text]
+
+-- | The cost model parameters are modelled coarsely as a set of named, integer parameters. Conversion to the
+-- "real" model type can fail if those don't line up with the parameters that we expect.
+type CostModelParameters = Map String Integer
+
+-- | Scripts to the ledger are serialised bytestrings.
+type Script = ShortByteString
+
+-- | Errors that can be thrown when evaluating a Plutus script.
+data EvaluationError =
+    CekError (UPLC.CekEvaluationException PLC.DefaultUni PLC.DefaultFun) -- ^ An error from the evaluator itself
+    | DeBruijnError PLC.FreeVariableError -- ^ An error in the pre-evaluation step of converting from de-Bruijn indices
+    | CodecError Flat.DecodeException -- ^ A serialisation error
+    | IncompatibleVersionError (PLC.Version ()) -- ^ An error indicating a version tag that we don't support
+    | CostModelParameterMismatch String -- ^ An error indicating that the cost model parameters didn't match what we expected
+    deriving stock (Show, Eq)
+
+instance Pretty EvaluationError where
+    pretty (CekError e)      = prettyClassicDef e
+    pretty (DeBruijnError e) = pretty e
+    pretty (CodecError e) = viaShow e
+    pretty (IncompatibleVersionError actual) = "This version of the Plutus Core interface does not support the version indicated by the AST:" <+> pretty actual
+    pretty (CostModelParameterMismatch e) = pretty e
+
+-- | Shared helper for the evaluation functions, deserializes the 'Script' , applies it to its arguments, and un-deBruijn-ifies it.
+mkTermToEvaluate :: (MonadError EvaluationError m) => Script -> [Data] -> m (UPLC.Term UPLC.Name PLC.DefaultUni PLC.DefaultFun ())
+mkTermToEvaluate bs args = do
+    (Scripts.Script (UPLC.Program _ v t)) <- liftEither $ first CodecError $ Flat.unflat $ fromShort bs
+    unless (v == PLC.defaultVersion ()) $ throwError $ IncompatibleVersionError v
+    let namedTerm = UPLC.termMapNames PLC.fakeNameDeBruijn t
+        -- This should go away when Data is a builtin
+        termArgs = fmap PlutusTx.lift args
+        applied = PLC.mkIterApp () namedTerm termArgs
+    liftEither $ first DeBruijnError $ PLC.runQuoteT $ UPLC.unDeBruijnTerm applied
+
+-- TODO: actually implement log output
+-- | Evaluates a script, with a budget that restricts how many resources it can use.
+evaluateScriptRestricting
+    :: VerboseMode -- ^ Whether to produce log output
```

Comment:
how do we ensure that VerboseMode on/off does not affect the execution units required to run the script? I forget. Is it just by not assigning any execution cost to logging inside a script?

Reply:
The cost of tracing primitives will not depend on `VerboseMode`, that would indeed be unsound.